### PR TITLE
PMM-8322: Delete artifact from storage

### PR DESCRIPTION
**Pull request by oter**
[PMM-8322](https://jira.percona.com/browse/PMM-8322)

- [x] Build: https://github.com/Percona-Lab/pmm-submodules/pull/1896

- [x] Links to other linked pull requests (optional).
---
- [x] Tests passed.
- [x] Feature build pass.
- [x] (Re)requested review.
- [x] Fix conflicts with target branch.
- [x] Update API dependency.
- [ ] Update jira ticket description if needed.

### CODE REVIEW

File: services/management/backup/artifacts_service.go [hunk 1]
```diff
@@ -96,6 +100,90 @@ func (s *ArtifactsService) ListArtifacts(context.Context, *backupv1beta1.ListArt
...
 }
 
+// DeleteArtifact deletes specified artifact.
+func (s *ArtifactsService) DeleteArtifact(
+	ctx context.Context,
+	req *backupv1beta1.DeleteArtifactRequest,
+) (*backupv1beta1.DeleteArtifactResponse, error) {
+	var s3Config *models.S3LocationConfig
+	var artifactName string
+	if err := s.db.InTransaction(func(tx *reform.TX) error {
```
Comment (Dasio) [inline services/management/backup/artifacts_service.go:110]: why are there 2 transactions? each one is finding artifact. Can't we merge it?

File: services/management/backup/artifacts_service.go [hunk 2]
```diff
@@ -96,6 +100,90 @@ func (s *ArtifactsService) ListArtifacts(context.Context, *backupv1beta1.ListArt
...
+			return err
+		}
+
+		artifactName = artifact.Name
+
+		location, err := models.FindBackupLocationByID(tx.Querier, artifact.LocationID)
+		if err != nil {
+			return err
+		}
+		if location.S3Config == nil {
```
Comment (Dasio) [inline services/management/backup/artifacts_service.go:126]: I would remove this condition. And later check if s3config is not nil, if not, then call s3 api.

File: services/management/backup/locations_service.go [hunk 3]
```diff
@@ -247,7 +247,7 @@ func (s *LocationsService) RemoveLocation(ctx context.Context, req *backupv1beta
 		mode = models.RemoveCascade
 	}
 	err := s.db.InTransaction(func(tx *reform.TX) error {
-		if err := models.RemoveBackupLocation(s.db.Querier, req.LocationId, mode); err != nil {
+		if err := models.RemoveBackupLocation(tx.Querier, req.LocationId, mode); err != nil {
```
[the remark this answers is not in the document]
Comment (oter) [reply services/management/backup/locations_service.go:250]: fixed

File: services/management/backup/artifacts_service.go [hunk 4]
```diff
@@ -96,6 +100,90 @@ func (s *ArtifactsService) ListArtifacts(context.Context, *backupv1beta1.ListArt
...
+			return err
+		}
+
+		for _, ri := range restoreItems {
+			if err := models.RemoveRestoreHistoryItem(tx.Querier, ri.ID); err != nil {
+				return err
+			}
+		}
+
+		if err := models.DeleteArtifact(tx.Querier, req.ArtifactId); err != nil {
```
[the remark this answers is not in the document]
Comment (oter) [reply services/management/backup/artifacts_service.go:176]: fixed

File: services/minio/service.go [hunk 5]
```diff
@@ -22,12 +22,22 @@ import (
...
 
 	"github.com/percona/pmm-managed/models"
 )
 
 // Service is wrapper around minio client.
 type Service struct {
+	l *logrus.Entry
+}
+
+func NewService(l *logrus.Entry) *Service {
```
[the remark this answers is not in the document]
Comment (oter) [reply services/minio/service.go:37]: fixed

File: services/management/backup/artifacts_service.go [hunk 6]
```diff
@@ -96,6 +100,90 @@ func (s *ArtifactsService) ListArtifacts(context.Context, *backupv1beta1.ListArt
 	}, nil
 }
 
+// DeleteArtifact deletes specified artifact.
+func (s *ArtifactsService) DeleteArtifact(
```
[the remark this answers is not in the document]
Comment (oter) [reply services/management/backup/artifacts_service.go:104]: fixed

File: services/management/backup/artifacts_service.go (hunk 1 above)
Comment (oter) [reply services/management/backup/artifacts_service.go:109]: fixed

File: main.go [hunk 7]
```diff
@@ -817,7 +817,7 @@ func main() {
 			rulesService:         rulesService,
 			jobsService:          jobsService,
 			versionServiceClient: versionService,
-			minio:                &minio.Service{},
+			minio:                minio.New(logrus.WithField("component", "minio-client")),
```
Comment (BupycHuk) [inline main.go:820]: Usually we don't pass logrus entry as parameter, we can create field in `New` method

File: models/artifact_helpers.go [hunk 8]
```diff
@@ -205,14 +205,14 @@ func CreateArtifact(q *reform.Querier, params CreateArtifactParams) (*Artifact,
 	return row, nil
 }
 
-// ChangeArtifactParams are params for changing existing artifact.
-type ChangeArtifactParams struct {
+// UpdateArtifactParams are params for changing existing artifact.
+type UpdateArtifactParams struct {
```
Comment (BupycHuk) [inline models/artifact_helpers.go:209]: What is the reason to change naming? We use `Change` and `Remove` naming everywhere.

File: services/management/backup/artifacts_service.go [hunk 9]
```diff
@@ -96,6 +100,104 @@ func (s *ArtifactsService) ListArtifacts(context.Context, *backupv1beta1.ListArt
...
+				return err
+			}
+		}
+
+		return models.DeleteArtifact(tx.Querier, req.ArtifactId)
+	}); err != nil {
+		return nil, err
+	}
+
+	return &backupv1beta1.DeleteArtifactResponse{}, nil
```
Comment (BupycHuk) [inline services/management/backup/artifacts_service.go:198]: Whole this stuff should be done in one transaction.

File: services/management/backup/artifacts_service.go [hunk 10]
```diff
@@ -96,6 +100,104 @@ func (s *ArtifactsService) ListArtifacts(context.Context, *backupv1beta1.ListArt
...
+		default:
+			return err
+		}
+		if artifact.Status != models.SuccessBackupStatus && artifact.Status != models.ErrorBackupStatus {
+			return status.Errorf(codes.FailedPrecondition, "Artifact with ID %q isn't in the final state.", artifactID)
+		}
+
+		artifactName = artifact.Name
+
+		restoreItems, err := models.FindRestoreHistoryItems(tx.Querier, &models.RestoreHistoryItemFilters{
```
Comment (BupycHuk) [inline services/management/backup/artifacts_service.go:123]: can we just query restore items with status == in progress

File: models/artifact_helpers.go (hunk 8 above)
Comment (artemgavrilov) [reply models/artifact_helpers.go:209]: > We Remove things we Add, and Delete things we Create.
> We Add pre-existing things (you can’t create email server, for example, you just add its configuration to PMM) and Create new things (we control everything about alerting rule, for example, so we create it)

That's how naming logic was described by AlekSi

File: services/checks/checks.go [hunk 11]
```diff
@@ -448,10 +448,10 @@ func (s *Service) ChangeInterval(params map[string]check.Interval) error {
 		// to check intervals in the DB so that they can be re-applied
 		// once the checks have been re-loaded on restarts.
 		e := s.db.InTransaction(func(tx *reform.TX) error {
-			cs, err := models.FindCheckSettingsByName(s.db.Querier, c.Name)
+			cs, err := models.FindCheckSettingsByName(tx.Querier, c.Name)
```
Comment (artemgavrilov) [inline services/checks/checks.go:451]: Epic!

Comment (oter) [reply services/checks/checks.go:451]: yeah 😄

File: main.go (hunk 7 above)
Comment (oter) [reply main.go:820]: done

File: services/management/backup/artifacts_service.go (hunk 9 above)
Comment (oter) [reply services/management/backup/artifacts_service.go:198]: not sure it's a good idea to put the s3 call into the db transaction.

File: services/minio/service.go [hunk 12]
```diff
@@ -48,6 +59,59 @@ func (s *Service) GetBucketLocation(ctx context.Context, endpoint, accessKey, se
...
+
+		if rerr != nil {
+			rerr = errors.Wrapf(rerr, "listing objects error: %s", err.Error())
+		} else {
+			rerr = errors.WithStack(err)
+		}
+	}()
+
+	var errorsEncountered bool
+	for rErr := range minioClient.RemoveObjects(ctx, bucketName, objectsCh, minio.RemoveObjectsOptions{}) {
```
[the remark this answers is not in the document]
Comment (artemgavrilov) [reply services/minio/service.go:103]: Minor: rErr can be perallocated

File: services/management/backup/artifacts_service.go [hunk 13]
```diff
@@ -96,6 +100,104 @@ func (s *ArtifactsService) ListArtifacts(context.Context, *backupv1beta1.ListArt
 	}, nil
 }
 
+func (s *ArtifactsService) deleteArtifactRequirements(
```
Comment (artemgavrilov) [inline services/management/backup/artifacts_service.go:103]: I didn't get what this method does.

File: services/minio/service.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (oter) [reply services/minio/service.go:103]: can you give me an example of the rErr preallocation?

File: services/management/backup/artifacts_service.go (hunk 13 above)
Comment (oter) [reply services/management/backup/artifacts_service.go:103]: let's have a call

File: services/minio/service.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (artemgavrilov) [reply services/minio/service.go:103]:
```
var rErr minio.RemoveObjectError
for rErr = range minioClient.RemoveObjects...
```

File: services/management/backup/artifacts_service.go (hunk 13 above)
Comment (oter) [reply services/management/backup/artifacts_service.go:103]: discussed, will improve naming

File: services/management/backup/artifacts_service.go (hunk 9 above)
Comment (oter) [reply services/management/backup/artifacts_service.go]: reworked

File: services/management/backup/artifacts_service.go (hunk 10 above)
Comment (oter) [reply services/management/backup/artifacts_service.go:125]: yes, we can, I will do

Comment (oter) [reply services/management/backup/artifacts_service.go:125]: status filter added

File: models/restore_history_helpers.go [hunk 14]
```diff
@@ -53,8 +56,15 @@ func FindRestoreHistoryItems(q *reform.Querier, filters *RestoreHistoryItemFilte
 		if _, err := FindArtifactByID(q, filters.ArtifactID); err != nil {
 			return nil, err
 		}
+
 		conditions = append(conditions, fmt.Sprintf("artifact_id = %s", q.Placeholder(idx)))
 		args = append(args, filters.ArtifactID)
+		idx++
+	}
+
+	if filters != nil && filters.Status != nil {
```
Comment (Dasio) [inline models/restore_history_helpers.go:65]: let's move it, to one `if filters != nil` or pass it as value not pointer?

File: services/management/backup/artifacts_service.go [hunk 15]
```diff
@@ -96,6 +100,119 @@ func (s *ArtifactsService) ListArtifacts(context.Context, *backupv1beta1.ListArt
...
+	if err := s.db.InTransaction(func(tx *reform.TX) error {
+		artifact, err := models.FindArtifactByID(tx.Querier, artifactID)
+		switch {
+		case err == nil:
+		case errors.Is(err, models.ErrNotFound):
+			return status.Errorf(codes.NotFound, "Artifact with ID %q not found.", artifactID)
+		default:
+			return err
+		}
+		if artifact.Status != models.SuccessBackupStatus &&
```
Comment (Dasio) [inline services/management/backup/artifacts_service.go:117]: switch on artifact status and on default return err?

File: services/management/backup/artifacts_service_test.go [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+
+	findArtifact := func(artifactID string) *backupv1beta1.Artifact {
+		artifacts, err := artifactsSvc.ListArtifacts(ctx, &backupv1beta1.ListArtifactsRequest{})
+		require.NoError(t, err)
+		require.NotNil(t, artifacts)
+
+		var artifact *backupv1beta1.Artifact
+		for _, a := range artifacts.Artifacts {
+			if a.ArtifactId == artifactID {
+				artifact = a
```
Comment (Dasio) [inline services/management/backup/artifacts_service_test.go:80]:
```suggestion
				return a
```

File: models/restore_history_helpers.go (hunk 14 above)
Comment (oter) [reply models/restore_history_helpers.go:65]: I've tried this approach, but then linter argues that inner statements are too deep 😄

File: services/management/backup/artifacts_service.go (hunk 15 above)
Comment (oter) [reply services/management/backup/artifacts_service.go:117]: done

File: services/management/backup/artifacts_service_test.go (hunk 16 above)
Comment (oter) [reply services/management/backup/artifacts_service_test.go:80]: done

File: services/management/backup/artifacts_service.go [hunk 17]
```diff
@@ -96,6 +100,121 @@ func (s *ArtifactsService) ListArtifacts(context.Context, *backupv1beta1.ListArt
...
+	if err := s.db.InTransaction(func(tx *reform.TX) error {
+		artifact, err := models.FindArtifactByID(tx.Querier, artifactID)
+		switch {
+		case err == nil:
+		case errors.Is(err, models.ErrNotFound):
+			return status.Errorf(codes.NotFound, "Artifact with ID %q not found.", artifactID)
+		default:
+			return err
+		}
+		switch artifact.Status {
```
[the remark this answers is not in the document]
Comment (oter) [reply services/management/backup/artifacts_service.go]: fixed

File: models/restore_history_helpers.go [hunk 18]
```diff
@@ -40,21 +42,31 @@ func FindRestoreHistoryItems(q *reform.Querier, filters *RestoreHistoryItemFilte
 	var args []interface{}
 
 	idx := 1
-	if filters != nil && filters.ServiceID != "" {
-		if _, err := FindServiceByID(q, filters.ServiceID); err != nil {
-			return nil, err
+	if filters != nil {
```
[the remark this answers is not in the document]
Comment (oter) [reply models/restore_history_helpers.go:45]: @Dasio 😄

File: models/restore_history_helpers.go (hunk 14 above)
Comment (oter) [reply models/restore_history_helpers.go]: pointer removed

**Review by Dasio — APPROVED** [review]

**Review by artemgavrilov — APPROVED** [review]

File: services/management/backup/artifacts_service.go [hunk 19]
```diff
@@ -141,6 +143,12 @@ func (s *ArtifactsService) deleteArtifactRequirements(
 
 		s3Config = location.S3Config
 
+		if _, err := models.UpdateArtifact(tx.Querier, artifactID, models.UpdateArtifactParams{
+			Status: models.DeletingBackupStatus.Pointer(),
```
[the remark this answers is not in the document]
Comment (oter) [reply services/management/backup/artifacts_service.go:163]: done

File: services/management/backup/artifacts_service.go [hunk 20]
```diff
@@ -100,7 +100,7 @@ func (s *ArtifactsService) ListArtifacts(context.Context, *backupv1beta1.ListArt
 	}, nil
 }
 
-func (s *ArtifactsService) deleteArtifactRequirements(
+func (s *ArtifactsService) beginDeletingArtifact(
```
[the remark this answers is not in the document]
Comment (oter) [reply services/management/backup/artifacts_service.go:129]: comment added

File: services/management/backup/artifacts_service_test.go [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+		Name:       "artifact_name",
+		Vendor:     "MySQL",
+		LocationID: locationRes.ID,
+		ServiceID:  *agent.ServiceID,
+		DataModel:  "physical",
+		Status:     models.SuccessBackupStatus,
+	})
+	require.NoError(t, err)
+
+	findArtifact := func(artifactID string) *backupv1beta1.Artifact {
```
[the remark this answers is not in the document]
Comment (oter) [reply services/management/backup/artifacts_service_test.go:72]: This is a service layer test for artifacts, so I need to test `ListArtifacts` handler.

[the remark this answers is not in the document]
Comment (BupycHuk) [reply services/management/backup/artifacts_service_test.go:72]: To test ListArtifacts we can make separate test which will be clear for other developers.

File: services/management/backup/artifacts_service_test.go [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+
+	_, err = artifactsSvc.DeleteArtifact(ctx, &backupv1beta1.DeleteArtifactRequest{
+		ArtifactId:  artifact.ID,
+		RemoveFiles: true,
+	})
+	require.EqualError(t, err, "failed to remove")
+
+	foundArtifact := findArtifact(artifact.ID)
+	require.NotNil(t, foundArtifact)
+	assert.Equal(t, foundArtifact.Status, backupv1beta1.BackupStatus_BACKUP_STATUS_FAILED_TO_DELETE)
```
[the remark this answers is not in the document]
Comment (oter) [reply services/management/backup/artifacts_service_test.go]: done

File: services/management/backup/artifacts_service_test.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (oter) [reply services/management/backup/artifacts_service_test.go]: done

[the remark this answers is not in the document]
Comment (BupycHuk) [reply services/management/backup/artifacts_service_test.go]: this one is not resolved

**Comment by artemgavrilov** [comment]
I'm resolving conflicts

File: models/artifact_helpers.go (hunk 8 above)
Comment (BupycHuk) [reply models/artifact_helpers.go:227]: TBH that sounds very interesting but, looks like it's more related to API level, not model level.

**Review by BupycHuk — APPROVED** [review]

File: models/artifact_helpers.go (hunk 8 above)
Comment (artemgavrilov) [reply models/artifact_helpers.go:227]: I think if we will apply different naming rules to API and models we will be completely confused 😄 

For me, it's ok to have any of these names.

**Pull request merged**